h/views/client.py

```diff
@@ -8,30 +8,96 @@
 from h import session as h_session
 from h.auth.tokens import generate_jwt
 from h.util.view import json_view
+from h import __version__
+
+
+def url_with_path(url):
+    if urlparse.urlparse(url).path == '':
+        return '{}/'.format(url)
+    else:
+        return url
 
 
-def render_app(request, extra=None):
-    """Render a page that serves a preconfigured annotation client."""
+def _app_html_context(assets_env, api_url, service_url, sentry_public_dsn,
+                      websocket_url, ga_client_tracking_id):
+    """
+    Returns a dict of asset URLs and contents used by the sidebar app
+    HTML tempate.
+    """
+
+    # the serviceUrl parameter must contain a path element
+    service_url = url_with_path(service_url)
+
+    app_config = {
+        'apiUrl': api_url,
+        'serviceUrl': service_url,
+        'release': __version__
+    }
+
+    if websocket_url:
+        app_config.update({
+            'websocketUrl': websocket_url,
+        })
+
+    if sentry_public_dsn:
+        app_config.update({
+            'raven': {
+                'dsn': sentry_public_dsn,
+                'release': __version__
+            }
+        })
+
+    if ga_client_tracking_id:
+        app_config.update({
+            'googleAnalytics': ga_client_tracking_id
+        })
+
+    app_css_urls = assets_env.urls('app_css')
+    app_js_urls = assets_env.urls('app_js')
+
+    return {
+        'app_config': json.dumps(app_config),
+        'app_css_urls': app_css_urls,
+        'app_js_urls': app_js_urls,
+    }
+
+
+@view_config(route_name='sidebar_app',
+             renderer='h:templates/app.html.jinja2')
+def sidebar_app(request, extra=None):
+    """
+    Return the HTML for the Hypothesis client's sidebar application.
+
+    :param extra: A dict of optional properties specifying link tags and meta
+                  attributes to be included on the page.
+    """
+
     client_sentry_dsn = request.registry.settings.get('h.client.sentry_dsn')
-    html = client.render_app_html(
-        assets_env=request.registry['assets_client_env'],
-        # FIXME: The '' here is to ensure this has a trailing slash. This seems
-        # rather messy, and is inconsistent with the rest of the application's
-        # URLs.
-        api_url=request.route_url('api.index'),
-        service_url=request.route_url('index'),
-        sentry_public_dsn=client_sentry_dsn,
-        websocket_url=request.registry.settings.get('h.websocket_url'),
-        ga_client_tracking_id=request.registry.settings.get('ga_client_tracking_id'),
-        extra=extra)
-    request.response.text = html
-    return request.response
+    assets_env = request.registry['assets_client_env']
+    api_url = request.route_url('api.index')
+    service_url = request.route_url('index')
+    sentry_public_dsn = client_sentry_dsn
+    websocket_url = request.registry.settings.get('h.websocket_url')
+    ga_client_tracking_id = request.registry.settings.get('ga_client_tracking_id')
```

Comment:
There's some duplication of this code for getting variables out of the `request` object (the `embed()` view below pulls out `assets_env` and `service_url` in the same way that this code does). I wonder if this block could be moved to a helper function that takes the `request` and returns a dict of all these?